platform/source/minimal-printf/mbed_printf_armlink_overrides.c

```diff
@@ -66,17 +66,12 @@ int $Sub$$__2snprintf(char *buffer, size_t length, const char *format, ...)
     return result;
 }
 
-int $Sub$$__2vprintf(char *buffer, const char *format, ...)
+int $Sub$$vprintf(const char *format, va_list arguments)
```

Comment:
Basic issue: isn't this now overriding `vprintf` twice? Isn't there already a definition of `$Sub$$vprintf` in mbed_printf_wrapper.c? That's how it looks to me reading it, but then it apparently isn't working?

Reply:
It is a little tricky.
The ARM compiler does not recognise `int SUB_VPRINTF(const char *format, va_list arguments){}` as a definition for `vprintf` in `mbed_printf_wrapper.c`. However, it recognises it when it is changed to `int $Sub$$vprintf(const char *format, va_list arguments){}`.

It seems the ARM compiler requires `$Sub$$` to be in the function signature when it is defined.
So all those `SUB_*` definitions in `mbed_printf_wrapper.c` are ignored for the ARM compiler. Only the definitions in `mbed_printf_armliink_overrides.c` count.

Reply:
`SUB_VPRINTF` is defined as `$Sub$$vprintf` in mbed_printf_wrapper.c, isn't it? I must be missing something.

Reply:
> `SUB_VPRINTF` is defined as `$Sub$$vprintf` in mbed_printf_wrapper.c, isn't it? I must be missing something.

You are right but I am not sure how the compiler handles `$Sub$$` if a pre-processor definition is also used. What I do know is that it does use the definition in `mbed_printf_wrapper.c`.

Reply:
Wouldn't expect any oddity. Compiler documentation simply says "The dollar character `$` is allowed in identifiers." Preprocessor should treat that the same as any other identifier character then.

Reply:
Working with Hugues, we've established that `mbed_printf_wrapper.c` is actually generating the `__wrap` functions rather than the `$Sub$$` functions because it tests like

    #if defined(__GNUC__)
    #elif defined(TOOLCHAIN_ARM) || defined(__ICCARM)

But the ARM compiler is in GCC-"compatible" mode, so is defining `__GNUC__`.

Suggest reversing those blocks.

Then `mbed_printf_wrapper.c` should work for the base functions, but `mbed_printf_armlink_overrides` needs to provide all the corresponding `__2` functions. That includes the `v` functions, as we've established that the compiler will use `__2vprintf` if it knows the format. (The `mbed_error_printf` doesn't know the format, so uses `vsnprintf`, but `__2vsnprintf` is still possible in principle).

It looks like `mbed_prinf_armlink_overrides` is missing `__2vsprintf`, `__2fprintf` and `__2vfprintf`.

I wonder if possibly you should eliminate the repetition and avoid inconsistency by putting all the definitions into a header file. Then do

    #if defined __ARMCC_VERSION || defined __ICCARM__
    /* Intercept both printf and __2printf using $Sub$$ */
    #define SUB_PRINTF $Sub$$printf
    #include "printf_wrappers.h"
    #undef SUB_PRINTF
    #define SUB_PRINTF $Sub$$__2printf
    #include "printf_wrappers.h"
    #elif defined __GNUC__
    /* Intercept both printf  using __wrap_ */
    #define SUB_PRINTF __wrap_printf
    #include " printf_wrappers.h"
    #endif

Reply:
Actually, even neater, do what `mbed_retarget.cpp` does to determine the prefix. You don't need one define per function, just a define to paste on the `$Sub$$`, `$Sub$$__2` or `__wrap_` prefix.

Reply:
[Done](https://github.com/ARMmbed/mbed-os/compare/fb075bcf31084175fb0bf8d2aebf672cd4262191..99cd2ce82c746e9ffe7cd64cc02fb674c6c7affa).